Define _WKWebsiteDataStoreDelegate callback methods for WindowProxy property access

bnham opened the pull request:
#### d7e49edab1122fd458f73d06b7973439ef915b40
<pre>
Define _WKWebsiteDataStoreDelegate callback methods for WindowProxy property access
<a href="https://bugs.webkit.org/show_bug.cgi?id=265408">https://bugs.webkit.org/show_bug.cgi?id=265408</a>
<a href="https://rdar.apple.com/problem/118852728">rdar://problem/118852728</a>

Reviewed by David Kilzer.

We want to define some _WKWebsiteDataStoreDelegate callbacks that notify the embedder that certain
properties on WindowProxy get accessed on windows associated with either side of a window.open
call. This patch only defines the delegate method and doesn&apos;t include the logic that actually calls
the method.

Basically, if a site opens another site via window.open, we will:

1. Call this new delegate method the moment that the `window.open` call occurs with
`WKWindowProxyPropertyInitialOpen`.

2. If either side of the WindowProxy object is used via the opener relationship, then we will call
the delegate method with one of `WKWindowProxyPropertyPostMessage`, `WKWindowProxyPropertyClosed`,
or `WKWindowProxyPropertyOther` depending on whether the `postMessage`, `closed`, or some other
property is called on the `WindowProxy` object respectively.

The `directly` parameter is true if this `openedRegistrableDomain` was the domain that actually
direclty accessed `property` on the `WindowProxy` instance. Otherwise, the `openedRegistrableDomain`
is only blamed for accessing the property because it exists in the navigation history of the frame.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreDelegate.h:

Canonical link: <a href="https://commits.webkit.org/271575@main">https://commits.webkit.org/271575@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f3961f5d0617c15a8ba39ed2a6aa0da0c1239d70

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28841 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7482 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30212 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31458 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/26308 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9617 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4842 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26354 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29111 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/6198 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/24774 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5388 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/5536 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/32796 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/26394 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/26226 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/31775 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/5508 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/3680 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/29556 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/7135 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6893 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/5985 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/6029 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreDelegate.h:
```diff
@@ -60,4 +67,5 @@ WK_API_AVAILABLE(macos(10.15), ios(13.0))
 - (void)websiteDataStore:(WKWebsiteDataStore *)dataStore navigateToNotificationActionURL:(NSURL *)url;
 - (void)requestBackgroundFetchPermission:(NSURL *)mainFrameURL frameOrigin:(NSURL *)frameURL  decisionHandler:(void (^)(bool isGranted))decisionHandler;
 - (void)notifyBackgroundFetchChange:(NSString *)backgroundFetchIdentifier change:(WKBackgroundFetchChange)change;
+- (void)websiteDataStore:(WKWebsiteDataStore *)dataStore domain:(NSString *)registrableDomain didOpenDomainViaWindowOpen:(NSString *)openedRegistrableDomain andAccessProperty:(WKWindowProxyProperty)property directly:(BOOL)directly;
```
I would change `andAccessProperty:` to `withAccessProperty:`.  It's very rare to use `and` in an Objective-C method signature.

bnham replied:
Maybe I'll just go with `withProperty:` then.

ddkilzer reviewed (approved):
r=me with a minor method name change.

bnham committed:
Define _WKWebsiteDataStoreDelegate callback methods for WindowProxy property access
https://bugs.webkit.org/show_bug.cgi?id=265408
rdar://problem/118852728

Reviewed by David Kilzer.

We want to define some _WKWebsiteDataStoreDelegate callbacks that notify the embedder that certain
properties on WindowProxy get accessed on windows associated with either side of a window.open
call. This patch only defines the delegate method and doesn't include the logic that actually calls
the method.

Basically, if a site opens another site via window.open, we will:

1. Call this new delegate method the moment that the `window.open` call occurs with
`WKWindowProxyPropertyInitialOpen`.

2. If either side of the WindowProxy object is used via the opener relationship, then we will call
the delegate method with one of `WKWindowProxyPropertyPostMessage`, `WKWindowProxyPropertyClosed`,
or `WKWindowProxyPropertyOther` depending on whether the `postMessage`, `closed`, or some other
property is called on the `WindowProxy` object respectively.

The `directly` parameter is true if this `openedRegistrableDomain` was the domain that actually
direclty accessed `property` on the `WindowProxy` instance. Otherwise, the `openedRegistrableDomain`
is only blamed for accessing the property because it exists in the navigation history of the frame.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreDelegate.h:

Canonical link: https://commits.webkit.org/271575@main
